### PR TITLE
Add Trainer support for ReduceLROnPlateau

### PR DESCRIPTION
**Pull request by pie3636**
# What does this PR do?

This PR solves #16503 by adding support to pytorch's [ReduceLROnPlateau](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html) to `Trainer`.

It does so by adding a new `REDUCE_ON_PLATEAU` field to `SchedulerType` and a new `reduce_lr_on_plateau_args` parameter to `TrainingArguments` that is parsed at initialization to avoid adding 9 new individual arguments. The scheduler re-uses the metric stored in `metric_for_best_model`, and is delayed to run after evaluation since it requires metrics to be populated.

I'm not sure whether it is due to the complexity of `Trainer`, my lack of experience (this is my first PR to a large project) or the uniqueness of `ReduceLROnPlateau` compared to other schedulers, but this PR feels a bit hacky, so I welcome any feedback.

## Before submitting
- [ ] This PR fixes a typo or improves the docs (you can dismiss the other checks if that's the case).
- [x] Did you read the [contributor guideline](https://github.com/huggingface/transformers/blob/main/CONTRIBUTING.md#start-contributing-pull-requests),
      Pull Request section?
- [x] Was this discussed/approved via a Github issue or the [forum](https://discuss.huggingface.co/)? Please add a link
      to it if that's the case.
- [x] Did you make sure to update the documentation with your changes? Here are the
      [documentation guidelines](https://github.com/huggingface/transformers/tree/main/docs), and
      [here are tips on formatting docstrings](https://github.com/huggingface/transformers/tree/main/docs#writing-source-documentation).
- [x] Did you write any new necessary tests?

## Who can review?

Anyone in the community is free to review the PR once the tests have passed. Feel free to tag
members/contributors who may be interested in your PR.

Looking at #16503, I believe this is for @sgugger.

### CODE REVIEW

File: src/transformers/trainer.py [hunk 1]
```diff
@@ -1204,6 +1205,36 @@ def get_optimizer_cls_and_kwargs(args: TrainingArguments) -> Tuple[Any, Any]:
...
+                    scheduler_args[key] = float(value)
+                elif key in ("patience", "cooldown"):
+                    scheduler_args[key] = int(value)
+                elif key == "min_lr":  # can be stored as a float or list of floats
+                    try:
+                        scheduler_args[key] = float(value)
+                    except ValueError:
+                        scheduler_args[key] = [float(x) for x in value[1:-1].split(",")]
+                else:
+                    scheduler_args[key] = value
```
Comment (sgugger) [inline src/transformers/trainer.py:1234]: Let's pick good defaults here and let users who want to customize them pass their own scheduler. The important bit is to support this scheduler (which you're doing).

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -723,6 +723,9 @@ class TrainingArguments:
     logging_first_step: bool = field(default=False, metadata={"help": "Log the first global_step"})
     logging_steps: int = field(default=500, metadata={"help": "Log every X updates steps."})
     logging_nan_inf_filter: bool = field(default=True, metadata={"help": "Filter nan and inf losses for logging."})
+    reduce_lr_on_plateau_args: Optional[str] = field(
+        default=None, metadata={"help": "Optional arguments to supply to ReduceLROnPlateau."}
+    )
```
Comment (sgugger) [inline src/transformers/training_args.py:728]: I would avoid adding such a specific arg. See comment above.

**Review by sgugger — COMMENTED** [review]
Thanks for your PR and adding tests! I'm a bit wary of adding yet another training argument just for this scheduler. How about we use good defaults and document that to customize them, users will need to pass their own scheduler to the `Trainer`? The important bits here are the lines you change in the `Trainer` to fully support this scheduler.

**Comment by pie3636** [comment]
Thanks for the review! I believe this should do it. There isn't much in the way of default arguments, but `ReduceLROnPlateau` is quite different from other schedulers in the first place.

**Review by sgugger — APPROVED** [review]
Thanks a lot!

**Comment by lombardata** [comment]
Hi @pie3636 and @sgugger , 
Thanks for this PR !
I would like to use the ReduceLROnPlateau scheduler but I don't understand which parameters (patience; factor; cooldown) it has by default and where I can change them.
If I'm right it uses the default ones :
`lr_scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, factor=0.2, patience=5, cooldown=2)`
and if I want to test new ones I have to build my own scheduler and pass it to hf?
Thanks a lot for this new feature !

**Comment by pie3636** [comment]
@lombardata The values you are referring to (`factor=0.2, patience=5, cooldown=2`) are example values used in the unit test.
The actual default parameters are [the ones provided by pytorch](https://pytorch.org/docs/stable/generated/torch.optim.lr_scheduler.ReduceLROnPlateau.html), which are, as I'm writing these lines, `optimizer, mode='min', factor=0.1, patience=10, threshold=0.0001, threshold_mode='rel', cooldown=0, min_lr=0, eps=1e-08, verbose=False`. 

If you want to use different values, you indeed need to build your own scheduler and optimizer, and pass them as a tuple to the `Trainer` class using the `optimizers` argument.

**Comment by lombardata** [comment]
Ok, thank you very much @pie3636 for the clarification.
If anyone read this post and is interested in how to pass a custom reducelronplateau scheduler to the Trainer, here is  a simple way to do it :
`optimizer = torch.optim.Adam(model.parameters(), lr = 0.01, weight_decay = 0.0001)
`
`lr_scheduler = torch.optim.lr_scheduler.ReduceLROnPlateau(optimizer, mode='min', factor=0.1, patience=5, verbose=True)`

```
trainer = Trainer(
    model=model,
    args=training_args,
    data_collator=collate_fn,
    train_dataset=prepared_ds["train"],
    eval_dataset=prepared_ds["validation"],
    tokenizer=feature_extractor,
    compute_metrics=compute_metrics,
    optimizers=(optimizer, lr_scheduler)
)

```
